kernel/src/log_segment.rs

```diff
@@ -309,75 +309,84 @@ fn list_log_files_with_version(
+                        file.file_type, version
+                    );
                 }
             }
 
-            // Group checkpoint parts by the number of parts they have
-            let mut checkpoints = HashMap::new();
-            for part_file in new_checkpoint_parts {
-                use LogPathFileType::*;
-                match &part_file.file_type {
-                    SinglePartCheckpoint
-                    | UuidCheckpoint(_)
-                    | MultiPartCheckpoint {
-                        part_num: 1,
-                        num_parts: 1,
-                    } => {
-                        // All single-file checkpoints are equivalent, just keep one
-                        checkpoints.insert(1, vec![part_file]);
-                    }
-                    MultiPartCheckpoint {
-                        part_num: 1,
-                        num_parts,
-                    } => {
-                        // Start a new multi-part checkpoint with at least 2 parts
-                        checkpoints.insert(*num_parts, vec![part_file]);
-                    }
-                    MultiPartCheckpoint {
-                        part_num,
-                        num_parts,
-                    } => {
-                        // Continue a new multi-part checkpoint with at least 2 parts
-                        if let Some(part_files) = checkpoints.get_mut(num_parts) {
-                            if *part_num == 1 + part_files.len() as u32 {
-                                // Safe to append because all previous parts exist
-                                part_files.push(part_file);
-                            }
-                        }
-                    }
-                    Commit | CompactedCommit { .. } | Unknown => {} // invalid file type => do nothing
-                }
-            }
-
-            // Find a complete checkpoint (all parts exist)
-            if let Some((_, complete_checkpoint)) = checkpoints
+            // Group and find the first complete checkpoint for this version.
+            // All checkpoints for the same version are equivalent, so we only take one.
+            if let Some((_, complete_checkpoint)) = group_checkpoint_parts(new_checkpoint_parts)
                 .into_iter()
-                .find(|(num_parts, part_files)| part_files.len() as u32 == *num_parts)
+                .find(|(num_parts, part_files)| part_files.len() == *num_parts as usize)
             {
-                // Validate the checkpoint before updating state
-                if validate_checkpoint_parts(version, &complete_checkpoint) {
-                    checkpoint_parts = complete_checkpoint;
-                    commit_files.clear(); // Clear commit files once checkpoint is found
-                }
+                checkpoint_parts = complete_checkpoint;
+                commit_files.clear(); // Clear commits when a complete checkpoint is found
```

Comment:
Again, better to doc "why" rather than "what"
```suggestion
                commit_files.clear(); // Log replay only needs commits after a complete checkpoint
```